src/osd/PG.cc

```diff
@@ -8558,7 +8558,7 @@ boost::statechart::result PG::RecoveryState::Active::react(const MNotifyRec& not
 		       << dendl;
     pg->proc_replica_info(
       notevt.from, notevt.notify.info, notevt.notify.epoch_sent);
-    if (pg->have_unfound()) {
+    if (pg->have_unfound() || pg->might_have_unfound.count(notevt.from)) {
```

Comment:
`WaitDeleteReserved` does not expect a `MQuery` event. so it crashed.
```
2019-04-04 05:36:29.570 7f2e3dc04700  7 osd.1 95 handle_fast_pg_query pg_query(1.2 epoch 95) v4 from osd.0
2019-04-04 05:36:29.570 7f2e3dc04700 15 osd.1 95 enqueue_peering_evt 1.2 epoch_sent: 95 epoch_requested: 95 MQuery 1.2 from 0 query_epoch 95 query: query(fulllog 0'0 epoch_sent 95)
2019-04-04 05:36:29.570 7f2e3dc04700 20 osd.1 op_wq(2) _enqueue OpQueueItem(1.2 PGPeeringEvent(epoch_sent: 95 epoch_requested: 95 MQuery 1.2 from 0 query_epoch 95 query: query(fulllog 0'0 epoch_sent 95)) prio 255 cost 10 e95)
2019-04-04 05:36:29.574 7f2e1ef57700 -1 /build/ceph-15.0.0-18-gc9721a2/src/osd/PG.cc: In function 'PG::RecoveryState::Crashed::Crashed(boost::statechart::state<PG::RecoveryState::Crashed, PG::RecoveryState::RecoveryMachine>::my_context)' thread 7f2e1ef57700 time 2019-04-04 05:36:29.571358
/build/ceph-15.0.0-18-gc9721a2/src/osd/PG.cc: 7039: abort()

 ceph version 15.0.0-18-gc9721a2 (c9721a21a2588c9131b0b1413a143aaa250943d1) octopus (dev)
 1: (ceph::__ceph_abort(char const*, int, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0xda) [0x84d8dd]
 2: (PG::RecoveryState::Crashed::Crashed(boost::statechart::state<PG::RecoveryState::Crashed, PG::RecoveryState::RecoveryMachine, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0>::my_context)+0xd9) [0xa50329]
 3: ceph-osd() [0xa6f096]
 4: (boost::statechart::simple_state<PG::RecoveryState::Started, PG::RecoveryState::RecoveryMachine, PG::RecoveryState::Start, (boost::statechart::history_mode)0>::react_impl(boost::statechart::event_base const&, void const*)+0x24b) [0xab19bb]
 5: (boost::statechart::simple_state<PG::RecoveryState::WaitDeleteReserved, PG::RecoveryState::ToDelete, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0>::react_impl(boost::statechart::event_base const&, void const*)+0xc1) [0xaac2a1]
 6: (boost::statechart::state_machine<PG::RecoveryState::RecoveryMachine, PG::RecoveryState::Initial, std::allocator<void>, boost::statechart::null_exception_translator>::process_event(boost::statechart::event_base const&)+0x6b) [0xa7f1cb]
 7: (PG::do_peering_event(std::shared_ptr<PGPeeringEvent>, PG::RecoveryCtx*)+0x122) [0xa6e902]
 8: (OSD::dequeue_peering_evt(OSDShard*, PG*, std::shared_ptr<PGPeeringEvent>, ThreadPool::TPHandle&)+0x1b4) [0x9a9174]
 9: (PGPeeringItem::run(OSD*, OSDShard*, boost::intrusive_ptr<PG>&, ThreadPool::TPHandle&)+0x50) [0xc2e7e0]
 10: (OSD::ShardedOpWQ::_process(unsigned int, ceph::heartbeat_handle_d*)+0xbeb) [0x99d2ab]
 11: (ShardedThreadPool::shardedthreadpool_worker(unsigned int)+0x4ac) [0xfc384c]
 12: (ShardedThreadPool::WorkThreadSharded::entry()+0x10) [0xfc6a00]
 13: (()+0x76ba) [0x7f2e402d16ba]
 14: (clone()+0x6d) [0x7f2e3f8d841d]
```

http://pulpito.ceph.com/kchai-2019-04-04_04:19:50-rados-wip-kefu-testing-2019-04-04-0949-distro-basic-mira/

Reply:
@neha-ojha 

Seems we still need that pg->is_degraded() checker. Because if pg is currently active+clean, `proc_replica_info` might have sent out a purge (MOSDPGRemove) message, and hence unconditionally sending out a follow-up MQuery (for Log) message will crash that stray peer.

Reply:
I'm of course no expert but regardless of `is_degraded`, sending out the message shouldn't be able to crash the peer, right?

Reply:
> @neha-ojha
> 
> Seems we still need that pg->is_degraded() checker. Because if pg is currently active+clean, `proc_replica_info` might have sent out a purge (MOSDPGRemove) message, and hence unconditionally sending out a follow-up MQuery (for Log) message will crash that stray peer.

@xiexingguo yeah, relying on `might_have_unfound` is not helpful in that scenario.

Reply:
> I'm of course no expert but regardless of `is_degraded`, sending out the message shouldn't be able to crash the peer, right?

That's because the state machines have been designed with some invariants in mind, which should not be broken, and if do break them, we end up in the `Crashed` state, like we did in http://pulpito.ceph.com/kchai-2019-04-04_04:19:50-rados-wip-kefu-testing-2019-04-04-0949-distro-basic-mira/3808022/. Let's use the `pg->is_degraded()` check.

Reply:
alright, back to my original patch :)